### PR TITLE
Fix nested context manager for main_process_first()

### PR DESCRIPTION
**Pull request by flukeskywalker**
This PR proposes a fix for #1303.

The bug seems to be in delegation from one context manager (for example from `Accelerator`) to another one (from `AccelerateState`). Currently, the following pattern is used:

```python
@contextmanager
def main_process_first(self):
    yield self.state.main_process_first()
```
However, when executed, this context manager simply yields without entering `self.state.main_process_first()`, causing the bug. The solution is to correctly enter the inner context manager using `with`.

This was a very interesting bug. Since I haven't defined nested context managers like this before, I wanted to understand what the right way to do this is, and whether the current implementation in the library was correct. Like the cool kids I even asked ChatGPT, which explained what the current implementation was trying to do and how to use `yield context_manager()` or `yield from context_manager()`. Unfortunately it turned out that ChatGPT was hallucinating its explanation and use of these concepts, and even the code examples which it produced did not run and produced an error instead. 

In summary, according to ChatGPT the current implementation is correct, though it's not. So I'm very curious if I'm understanding it correctly now, and whether GPT-3/Copilot/ChatGPT had anything to do with the original implementation :)

### CODE REVIEW

**Comment by muellerzr** [comment]
Hi @flukeskywalker, it did not actually. But great job identifying the bug! What concerns me is when I was writing this I tried my best to come up with a test, and clearly that test does not work. So if you feel inclined to do so, it'd be great to see perhaps what went wrong with them/fix them (see accelerate/test_utils/scripts/test_script.py), otherwise I will look into it later next week.

**Comment by flukeskywalker** [comment]
Thanks @muellerzr. Indeed, the current test doesn't work because the `process_index` does not depend on whether the main process executes the code block first or not. 

We need a way to test not only that the code block is entered first in the main process, but also that it is exited in the main process before it is entered in other processes. This will ensure that there isn't a false positive when the main process happened to enter the code block first. A simple test would be to create a file if in main process and attempt to read it if not in main process. Perhaps there could be something more elegant?

**Comment by muellerzr** [comment]
I think that test would work quite well for the time being!

**Comment by muellerzr** [comment]
@flukeskywalker would you like to include it in this PR? Otherwise I can write one in afterwards

**Comment by flukeskywalker** [comment]
Added, ready for review.

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -755,7 +755,8 @@ def main_process_first(self):
         ...     print(f"This will be printed by process {accelerator.process_index}")
         ```
         """
-        yield self.state.main_process_first()
```
Comment (lebrice) [inline src/accelerate/accelerator.py:758]: Hey there, just a thought: you also could have written this as `yield from self.state.main_process_first()`.
I personally prefer your way of doing it, it feels a bit more explicit.

**Review by muellerzr — APPROVED** [review]
Excellent work! (Sorry it took a few days, was on leave). 

Just a small nit/idea for the test to make it a bit more rigid. Otherwise great 🤗

File: src/accelerate/accelerator.py (hunk 1 above)
Comment (muellerzr) [reply src/accelerate/accelerator.py:758]: Agreed, explicit is great here. Especially with something advanced such as this

File: src/accelerate/test_utils/scripts/test_script.py [hunk 2]
```diff
@@ -55,10 +57,13 @@ def process_execution_check():
...
-    idxs = accelerator.gather(idx)
-    if num_processes > 1:
-        assert idxs[0] == 0, "Main process was not first."
+        if accelerator.is_main_process:
+            time.sleep(0.1)  # ensure main process takes longest
+            open("check_main_process_first.txt", "w").close()
+        assert os.path.isfile("check_main_process_first.txt"), "Main process did not create check file yet."
+    accelerator.wait_for_everyone()
+    if accelerator.is_main_process:
+        os.remove("check_main_process_first.txt")
```
[the remark this answers is not in the document]
Comment (flukeskywalker) [reply src/accelerate/test_utils/scripts/test_script.py:66]: Thanks! I like the expansion idea. I should point out a couple of issues though:
- `PartialState().main_process_first()` actually works fine in the current release. The issue is in the delegation of the context manager in `Accelerator` to that in `State` and so on.
- If we switch to `Accelerator` the test doesn't do what we want because the main process writes first due to the time delays, causing the test to succeed with the current release. This is why I had a delay in the main process _before_ creating the file and no delay in other processes, to ensure that the full block of code in the context must be executed on main process before being entered in other processes. I hope I'm thinking correctly; does that makes sense? :)

I'll try to modify the test to incorporate this idea tomorrow.

File: src/accelerate/accelerator.py (hunk 1 above)
Comment (flukeskywalker) [reply src/accelerate/accelerator.py:758]: Hi @lebrice!

Interestingly, I think what you suggest doesn't work, because `State.main_process_first()` is not a generator but another context manager. BTW this is why `PartialState.main_process_first()` does work -- it delegates to a generator.

This is related to what I found very interesting in the issue as I mentioned above: it _feels like_ `yield` or at least `yield from` should work here, but it doesn't. I think this is why ChatGPT is confident it works :D It suggested the following example should work, which actually produces an error (you can remove the decorator from  `inner_context_manager()` to make it work).

```python
from contextlib import contextmanager

@contextmanager
def outer_context_manager():
    print("Entering outer context")
    yield
    print("Exiting outer context")

@contextmanager
def inner_context_manager():
    print("Entering inner context")
    yield
    print("Exiting inner context")

@contextmanager
def combined_context_manager():
    print("Entering combined context")
    yield from inner_context_manager()
    print("Exiting combined context")

with outer_context_manager():
    with combined_context_manager():
        print("Inside combined context")
```

File: src/accelerate/test_utils/scripts/test_script.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (flukeskywalker) [reply src/accelerate/test_utils/scripts/test_script.py]: Modified the test to incorporate the suggestion. Confirmed that it currently fails on 0.18.0 and passes with the fix. Should be ready for review.

**Comment by muellerzr** [comment]
@flukeskywalker do `make style; make quality` :)

**Comment by flukeskywalker** [comment]
Hi @muellerzr, do you think this PR is ready?

**Review by muellerzr — APPROVED** [review]
Yep! Thanks for the fix and iterating with it!
